src/Types/ListType.php

```diff
@@ -0,0 +1,104 @@
+<?php
+
+namespace ILIAS\Types;
+
+/**
+ * Class ListType
+ *
+ * @package ILIAS\Types
+ *
+ * @author  Oskar Truffer <ot@studer-raimann.ch>
+ *
+ * Describes a list of a certain Type.
+ *
+ * Attention:
+ * isExtensionOf behaves Covariant!
+ * new ListType(Dog::class).isExtensionOf(new ListType(Animal::class)) == true!
+ *
+ * See: https://en.wikipedia.org/wiki/Covariance_and_contravariance_(computer_science)
+ */
+class ListType implements Type, Ancestors {
+
+	/** @var Type */
+	protected $type;
+
+
+	/**
+	 * SingleType constructor.
+	 *
+	 * @param $fullyQualifiedClassName string|Type Give a Value Type or a Type that will be wrapped in a single type.
+	 */
+	public function __construct($fullyQualifiedClassName) {
+		if (!is_a($fullyQualifiedClassName, Type::class)) {
+			$fullyQualifiedClassName = new SingleType($fullyQualifiedClassName);
+		}
+		$this->type = $fullyQualifiedClassName;
+	}
+
+
+	/**
+	 * @inheritdoc
+	 */
+	function __toString() {
+		return "[" . $this->type . "]";
+	}
+
+
+	/**
+	 * Is this type a subtype of $type. Not strict! x->isExtensionOf(x) == true.
+	 *
+	 * Attention: [Dog].isExtensionOf([Animal]) == true. In other words:
+	 * The isExtensionOf behaves Covariant!
+	 *
+	 * If you are familiar with e.g. Java Generics ?.isExtensionOf(x) behaves the same as <? extends x>.
```

Comment:
Hi Oskar,
I really see your use case but I still disagree with the implementation. Let's try again =)

I think the problem boils down to these two problems:
* If I have any two types `A` and `B` the only things I could do with them are defined via the interface `Type`. If `A.isExtensionOf(B) === true`: What does it tell me? How does it help? I certainly cannot treat the `A` like a `B`, because this exactly breaks down in the list case. Thus, if I have a background task (or other thingy) acting on a `B`, I cannot feed an `A` into it.
* What is your understanding of covariant and contravariant? As far as I understand, List is neither covariant nor contravariant, hence the problem. `pop` is covariant. If I have a list that contains `A`s and pop values from it, I could treat the `A`s as `B`s if `A.isExtensionOf(B)`. `push` on the other hand is contravariant. If I have a List of `B`s I could push `A`s into it if `A.isExtensionOf(B)`, which is exactly the other way round. If I had a list of `A`s I cannot push `B`s into it, even if `A.isExtensionOf(B)`, because an `A` is somehow more than a `B`.

My goal is not to be in sync with Java, though, because Java does it wrong, as far as I understand, and is unsound there.

The idea with the `Extends`-type seems to be sound, though.

And also nice to have a discussion on this level here.

Best Regards!